File: src/languages/wu/wu.ts
/**
 * Wu Chinese / Shanghainese (wuu) phonemizer — canonical IPA. The third Sinitic language (after Mandarin and
 * Cantonese), and phonologically the most distinct: a THREE-WAY obstruent contrast that keeps the Middle Chinese
 * VOICED series (b d ɡ z d͡ʑ ʑ v ɦ — the yang register), glottalised yin sonorants (ʔm ʔn ʔl ʔŋ), a checked
 * glottal-stop coda (入声 → ʔ), front-rounded ø/y, the apical ɿ, and a LEFT-PROMINENT tone sandhi where the whole
 * prosodic word's melody is set by the first syllable's register.
 *
 * Written in Han characters; the front-end maps Han → Wugniu (zaonhe romanization) via dict.tsv (word→reading)
 * with greedy longest-match segmentation, so multi-char words carry their baked sandhi melody (上海 zaon2 he4)
 * and unknown chars fall back to per-char citation tone. The back-end (wu.jsonc) maps each Wugniu syllable →
 * IPA: [initial] + final + one of the tones as Chao contour letters. The dict is the rime-wugniu zaonhe schema;
 * the Wugniu→IPA back-end is authored here.
 */
import type { Phonemizer } from "../../registry.ts";
import { assembleClauses, clauseSink } from "../../core/clauses.ts";
import { LATIN_RUN } from "../../core/hostWord.ts";
import { loadManifest } from "../../core/loadManifest.ts";
import { loadTsvMap } from "../../core/loadTsv.ts";

interface WuDef {
    initials: Record<string, string>;
    finals: Record<string, string>;
    tones: Record<string, string>;
    clausePunctuation: Record<string, string>;
}
const DEF = loadManifest<WuDef>(import.meta.url, "wu.jsonc");
const CLAUSE_MARK = DEF.clausePunctuation;
// Initials tried longest-first so digraphs win (tsh>ts>t, ngh>ng>n>g, gh>g, sh>s …).
const INITIALS = Object.keys(DEF.initials).sort((a, b) => b.length - a.length);
const VOWEL = "aeiouy";
const SYLLABIC: Record<string, string> = { m: "m̩", n: "n̩", ng: "ŋ̍" };

let DICT: Map<string, string> | undefined;
function dict(): Map<string, string> {
    return (DICT ??= loadTsvMap(import.meta.url, "dict.tsv"));
}
const MAX_WORD = 8; // greedy segmentation window

const HAN = /\p{Script=Han}/u;
const WUGNIU = /^[a-z]+[0-9](?:\s+[a-z]+[0-9])*$/i;

/** One Wugniu syllable (e.g. "zaon2", "koq7") → IPA (initial + final + Chao tone). */
function syllableToIpa(syl: string): string {
    const m = /^([a-z]+?)([0-9])?$/i.exec(syl);
    if (!m) return syl;
    let body = m[1]!.toLowerCase();
    const tone = m[2] !== undefined ? (DEF.tones[m[2]] ?? "") : "";
    // Consonant onset (longest match) + final.
    for (const ini of INITIALS) {
        if (!body.startsWith(ini)) continue;
        const rime = body.slice(ini.length);
        const final = DEF.finals[rime];
        if (final !== undefined) return DEF.initials[ini]! + final + tone;
    }
    // Zero-onset: a whole-body final first (covers front-rounded yu/yun/yuq and the apical bare y), then the
    // glide onsets ⟨y⟩/⟨w⟩ + vowel (yi→ji, wa→wa), then a syllabic nasal.
    if (DEF.finals[body] !== undefined) return DEF.finals[body]! + tone;
    if ((body[0] === "y" || body[0] === "w") && VOWEL.includes(body[1] ?? "")) {
        const rest = DEF.finals[body.slice(1)];
        if (rest !== undefined) return (body[0] === "y" ? "j" : "w") + rest + tone;
    }
    if (SYLLABIC[body] !== undefined) return SYLLABIC[body]! + tone;
    return syl; // unknown rime → leave the romanization visible
}

/** A space-separated Wugniu reading → IPA. */
function wugniuToIpa(reading: string): string {
    return reading.trim().split(/\s+/u).map(syllableToIpa).join(" ");
}

/** A Han run → IPA (greedy longest-match over the dictionary; unknown chars are skipped). */
function hanRun(run: string): string {
    const chars = [...run];
    const out: string[] = [];
    for (let i = 0; i < chars.length; ) {
        let matched = "";
        let reading = "";
        for (let len = Math.min(MAX_WORD, chars.length - i); len >= 1; len--) {
            const word = chars.slice(i, i + len).join("");
            const hit = dict().get(word);
            if (hit) {
                matched = word;
                reading = hit;
                break;
            }
        }
        if (reading) {
            out.push(wugniuToIpa(reading));
            i += [...matched].length;
        } else i++; // no reading for this char → skip
    }
    return out.join(" ");
}

// Han numeral composition (shared Chinese system) — the Han string is then read through the dict, so number
// words pick up their sandhi and no separate number IPA is authored.
const DIGITS = ["零", "一", "二", "三", "四", "五", "六", "七", "八", "九"];
const SMALL = ["", "十", "百", "千"];
function under10000(n: number): string {
    if (n === 0) return "";
    let out = "";
    let zero = false;
    for (let p = 3; p >= 0; p--) {
        const unit = Math.floor(n / 10 ** p) % 10;
        if (unit === 0) {
            if (out) zero = true;
        } else {
            if (zero) out += DIGITS[0];
            zero = false;
            out += (p === 1 && unit === 1 && !out ? "" : DIGITS[unit]!) + SMALL[p]!;
        }
    }
    return out;
}
function integerToHan(n: number): string {
    if (n === 0) return DIGITS[0]!;
    if (n < 0) return "";
    const yi = Math.floor(n / 1_0000_0000);
    const wan = Math.floor((n % 1_0000_0000) / 10000);
    const rest = n % 10000;
    let out = "";
    if (yi) out += integerToHan(yi) + "億";
    if (wan) out += under10000(wan) + "萬";
    if (rest) {
        if ((yi || wan) && rest < 1000) out += DIGITS[0];
        out += under10000(rest);
    }
    return out;
}

export type ForeignPhonemizer = (latin: string) => string;

class WuPhonemizer implements Phonemizer {
    constructor(private foreign?: ForeignPhonemizer) {}
    text(input: string): string {
        // Whole-string Wugniu input (tone digits present) → direct path.
        if (WUGNIU.test(input.trim())) return wugniuToIpa(input);
        // `assembleClauses` rather than a private exec loop: this loop was already exactly that shape
        // (clauseSink + iterate the token regex), it just predated the shared helper — so it never got the
        // GAP PASS and a run in a script it does not own was dropped. The engine still claims Latin
        // itself; the gap pass covers everything else via the script router (core/scripts.ts).
                // ⚠ The Latin arm is ALL of Latin plus marks, not `[A-Za-z]+`: the ASCII class ended the token at a
                // diacritic and left that letter read as an English letter name (`Cañitas` → *kʰˈʌ ˈɛn ˈaᶦt̬əz*). The
                // engine routes a Latin run to the injected reader, so widening is the whole fix.
                const tok = new RegExp(`(\\p{Script=Han}+)|(\\d+)|(${LATIN_RUN})|([。，、？！；：.,?!;:])`, "gu");
        
        return assembleClauses(input, tok, (m, sink) => {
            if (m[1]) sink.emit(hanRun(m[1]));
            else if (m[2]) {
                const n = Number(m[2]);
                if (Number.isSafeInteger(n)) sink.emit(hanRun(integerToHan(n)));
            } else if (m[3]) sink.emit(this.foreign ? this.foreign(m[3]) : "");
            else if (m[4]) {
                const mk = CLAUSE_MARK[m[4]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Wu Chinese phonemizer. `foreign` handles embedded Latin runs. */
export function createWu(foreign?: ForeignPhonemizer): Phonemizer {
    return new WuPhonemizer(foreign);
}

/** Bare word→IPA (tests / eval): Han → IPA, or a direct Wugniu reading. */
export function phonemizeWord(word: string): string {
    return HAN.test(word) ? hanRun(word) : wugniuToIpa(word);
}

File: csharp/Vernacula.Phonemizer/Languages/Wu/Wu.cs
/**
 * Wu Chinese / Shanghainese (wuu) phonemizer — canonical IPA. Han → Wugniu (zaonhe romanization) via
 * dict.tsv with greedy longest-match segmentation, then Wugniu → [initial] + final + Chao tone letters.
 * Ported from src/languages/wu/wu.ts — see that file for the corpus evidence.
 */
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Wu;

public sealed class WuDef
{
    public IReadOnlyDictionary<string, string> Initials { get; init; } = new Dictionary<string, string>();
    public IReadOnlyDictionary<string, string> Finals { get; init; } = new Dictionary<string, string>();
    public IReadOnlyDictionary<string, string> Tones { get; init; } = new Dictionary<string, string>();
    public IReadOnlyDictionary<string, string> ClausePunctuation { get; init; } = new Dictionary<string, string>();
    public string MeasureWords { get; init; } = "";
    public IReadOnlyDictionary<string, string> LetterNames { get; init; } = new Dictionary<string, string>();
}

public static class WuPhonemizer
{
    public static readonly WuDef DEF = LoadManifest.Load<WuDef>("languages/wu", "wu.jsonc");
    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => DEF.ClausePunctuation;
    // Initials tried longest-first so digraphs win (tsh>ts>t, ngh>ng>n>g, gh>g, sh>s …). LINQ's
    // OrderByDescending is stable, as JS `Array.prototype.sort` is, so equal-length keys keep manifest order.
    private static readonly IReadOnlyList<string> INITIALS = DEF.Initials.Keys.OrderByDescending(k => k.Length).ToList();
    private const string VOWEL = "aeiouy";

    private static readonly IReadOnlyDictionary<string, string> SYLLABIC = new Dictionary<string, string>(StringComparer.Ordinal)
    {
        ["m"] = "m̩", ["n"] = "n̩", ["ng"] = "ŋ̍", ["mh"] = "ʔm̩", ["nh"] = "ʔn̩", ["ngh"] = "ʔŋ̍",
    };

    private static Dictionary<string, string>? DICT;
    private static readonly object GATE = new();
    private static Dictionary<string, string> Dict()
    {
        lock (GATE) return DICT ??= LoadTsv.LoadTsvMap("languages/wu", "dict.tsv");
    }
    private const int MAX_WORD = 8; // greedy segmentation window

    private static readonly JsRe HAN = JsRegex.Compile("\\p{Script=Han}", "u");
    private static readonly JsRe WUGNIU = JsRegex.Compile("^[a-z]+[0-9](?:\\s+[a-z]+[0-9])*$", "u");
    private static readonly JsRe SYLLABLE = JsRegex.Compile("^([a-z]+?)([0-9])?$", "i");
    private static readonly JsRe WHITESPACE = JsRegex.Compile("\\s+", "u");

    /** One Wugniu syllable (e.g. "zaon2", "koq7") → IPA (initial + final + Chao tone). */
    private static string SyllableToIpa(string syl)
    {
        var m = SYLLABLE.Match(syl);
        if (!m.Success) return syl;
        var body = m.Groups[1].Value.ToLowerInvariant();
        var tone = m.Groups[2].Success ? DEF.Tones.GetValueOrDefault(m.Groups[2].Value) ?? "" : "";
        foreach (var ini in INITIALS)
        {
            if (!body.StartsWith(ini, StringComparison.Ordinal)) continue;
            if (DEF.Finals.TryGetValue(body[ini.Length..], out var final)) return DEF.Initials[ini] + final + tone;
        }
        if (DEF.Finals.TryGetValue(body, out var whole)) return whole + tone;
        // ⚠ JS `VOWEL.includes(body[1] ?? "")` is TRUE for a one-letter body — `"".includes` of the empty
        // string. Reproduced: the branch is then harmless because `finals[""]` is undefined either way.
        var second = body.Length > 1 ? body[1].ToString() : "";
        if ((body.Length > 0 && (body[0] == 'y' || body[0] == 'w')) && VOWEL.Contains(second, StringComparison.Ordinal))
            if (DEF.Finals.TryGetValue(body[1..], out var rest))
                return (body[0] == 'y' ? "j" : "w") + rest + tone;
        if (SYLLABIC.TryGetValue(body, out var syllabic)) return syllabic + tone;
        return syl; // unknown rime → leave the romanization visible
    }

    /** A space-separated Wugniu reading → IPA. */
    private static string WugniuToIpa(string reading) =>
        string.Join(" ", SplitWs(Js.Trim(reading)).Select(SyllableToIpa));

    /** JS `s.split(/\s+/u)` — a leading empty piece is possible and is preserved, as there. */
    private static List<string> SplitWs(string s)
    {
        var parts = new List<string>();
        var last = 0;
        foreach (var m in JsRegex.MatchAll(WHITESPACE, s))
        {
            parts.Add(s[last..m.Index]);
            last = m.Index + m.Length;
        }
        parts.Add(s[last..]);
        return parts;
    }

    /** A Han run → IPA (greedy longest-match over the dictionary; unknown chars are skipped). */
    private static string HanRun(string run)
    {
        // UTF-16 offset of each code point, so slicing the window is O(1) rather than rebuilding a
        // code-point list per candidate. Indices are code points, exactly as the TS `[...run]` array is.
        var offs = new List<int>();
        for (var k = 0; k < run.Length; k += char.IsHighSurrogate(run[k]) && k + 1 < run.Length && char.IsLowSurrogate(run[k + 1]) ? 2 : 1)
            offs.Add(k);
        offs.Add(run.Length);
        var count = offs.Count - 1;

        var outp = new List<string>();
        for (var i = 0; i < count;)
        {
            var matchedLen = 0;
            var reading = "";
            for (var len = Math.Min(MAX_WORD, count - i); len >= 1; len--)
            {
                var word = run[offs[i]..offs[i + len]];
                if (Dict().TryGetValue(word, out var hit) && hit != "")
                {
                    matchedLen = len;
                    reading = hit;
                    break;
                }
            }
            if (reading != "")
            {
                outp.Add(WugniuToIpa(reading));
                i += matchedLen;
            }
            else i++; // no reading for this char → skip
        }
        return string.Join(" ", outp);
    }

    private static readonly string[] SMALL = { "", "十", "百", "千" };
    private static IReadOnlyList<string> DIGITS => Normalize.DIGITS;

    private static string Under10000(double n)
    {
        if (n == 0) return "";
        var outp = "";
        var zero = false;
        for (var p = 3; p >= 0; p--)
        {
            var unit = Math.Floor(n / Math.Pow(10, p)) % 10;
            if (unit == 0)
            {
                if (outp != "") zero = true;
            }
            else
            {
                if (zero) outp += DIGITS[0];
                zero = false;
                outp += (p == 1 && unit == 1 && outp == "" ? "" : DIGITS[(int)unit]) + SMALL[p];
            }
        }
        return outp;
    }

    private static string IntegerToHan(double n)
    {
        if (n == 0) return DIGITS[0];
        if (n < 0) return "";
        var yi = Math.Floor(n / 1_0000_0000);
        var wan = Math.Floor(n % 1_0000_0000 / 10000);
        var rest = n % 10000;
        var outp = "";
        if (yi != 0) outp += IntegerToHan(yi) + "億";
        if (wan != 0) outp += Under10000(wan) + "萬";
        if (rest != 0)
        {
            if ((yi != 0 || wan != 0) && rest < 1000) outp += DIGITS[0];
            outp += Under10000(rest);
        }
        return outp;
    }

    private static readonly JsRe TOKEN = JsRegex.Compile(
        $"(\\p{{Script=Han}}+)|(\\d+)|({HostWord.LATIN_RUN})|([。，、？！；：.,?!;:])", "gu");

    private sealed class Engine : ILanguage
    {
        private readonly Func<string, string>? _foreign;
        internal Engine(Func<string, string>? foreign) => _foreign = foreign;

        public string Text(string input)
        {
            // Whole-string Wugniu input (tone digits present) → direct path, BEFORE normalization.
            if (WUGNIU.IsMatch(Js.Trim(input))) return WugniuToIpa(input);
            input = Normalize.NormalizeWu(input, DEF.MeasureWords, DEF.LetterNames);

            return Clauses.AssembleClauses(input, TOKEN, (m, sink) =>
            {
                if (m.Groups[1].Success && m.Groups[1].Value.Length > 0) sink.Emit(HanRun(m.Groups[1].Value));
                else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0)
                {
                    var n = Js.Number(m.Groups[2].Value);
                    sink.Emit(HanRun(double.IsInteger(n) && Math.Abs(n) <= 9007199254740991d
                        ? IntegerToHan(n)
                        : Sinitic.SpellHanDigits(m.Groups[2].Value, DIGITS)));
                }
                else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
                    sink.Emit(_foreign is not null ? _foreign(m.Groups[3].Value) : "");
                else if (m.Groups[4].Success && m.Groups[4].Value.Length > 0)
                {
                    var mk = CLAUSE_MARK.GetValueOrDefault(m.Groups[4].Value);
                    if (mk is not null) sink.Pause(mk);
                }
            });
        }
    }

    /** Build the Wu Chinese phonemizer. `foreign` handles embedded Latin runs. */
    public static ILanguage CreateWu(Func<string, string>? foreign = null) => new Engine(foreign);

    /** Bare word→IPA (tests / eval): Han → IPA, or a direct Wugniu reading. */
    public static string PhonemizeWord(string word) => HAN.IsMatch(word) ? HanRun(word) : WugniuToIpa(word);

    internal static void RegisterSelf() =>
        Registry.Register("wu", () => CreateWu(Registry.ReadAsEnglish));
}
